### PR TITLE
[DNS] Fix DNS import/export issues.

### LINKED ISSUE #5259: DNS exports are relative, but include full domain name (missing a dot or omit full domain)
It looks like that **az network dns zone export** is exporting a zone with each record appended by the zone name, however not ending with a . (signalling it's an absolute/fully qualified entry). 

Using the example from pull request #2040, if i exported the DNS record for mycname in zone1.com, this is the exported result ($ORIGIN zone1.com.): 
`mycname.zone1.com 3600 IN CNAME contoso.com.`

However, missing the dot this could mean that it is treated as a relative record, and will then - fully qualified - be interpreted as mycname.zone1.com.zone1.com. This is also true when I import that same zone file/record with **az network dns zone import** (it will create a new record instead of updating an existing one): 
`mycname.zone1.com.zone1.com 3600 IN CNAME contoso.com.`

In that sense, there might be two exports that should/might work (not sure about the exact interpretation rules within the az command): 
fully qualified: `mycname.zone1.com. 3600 IN CNAME contoso.com.`
relative to $ORIGIN: `mycname 3600 IN CNAME contoso.com.`

---

### Environment summary

Azure CLI 2.0 on Windows 10.

### LINKED ISSUE #5362: [DNS] Importing a dns zone adds the tld twice to A records
Importing a dns zone adds the tld twice to A records. Expectations are that the fqdn would not be added to the record name.
---

### Environment summary

The issue only reproduces on 2.0.25

```
~$ az --version
azure-cli (2.0.25)

acr (2.0.19)
acs (2.0.24)
advisor (0.1.1)
appservice (0.1.24)
backup (1.0.6)
batch (3.1.8)
batchai (0.1.4)
billing (0.1.7)
cdn (0.0.11)
cloud (2.0.11)
cognitiveservices (0.1.10)
command-modules-nspkg (2.0.1)
configure (2.0.13)
consumption (0.2.1)
container (0.1.16)
core (2.0.25)
cosmosdb (0.1.16)
dla (0.0.17)
dls (0.0.19)
eventgrid (0.1.8)
extension (0.0.7)
feedback (2.0.8)
find (0.2.8)
interactive (0.3.13)
iot (0.1.16)
keyvault (2.0.16)
lab (0.0.15)
monitor (0.1.0)
network (2.0.21)
nspkg (3.0.1)
profile (2.0.17)
rdbms (0.0.11)
redis (0.2.11)
reservations (0.1.1)
resource (2.0.21)
role (2.0.17)
servicefabric (0.0.9)
sql (2.0.19)
storage (2.0.23)
vm (2.0.24)

Python location '/home/pjperez/lib/azure-cli/bin/python'
Extensions directory '/home/pjperez/.azure/cliextensions'

Python (Linux) 2.7.12 (default, Nov 19 2016, 06:48:10)
[GCC 5.4.0 20160609]

Legal docs and information: aka.ms/AzureCliLegal
```

When importing a simple zone file, A record "www" gets added as "www.domain.com.domain.com"

File:
```
; Exported zone file from Azure DNS
; Resource Group Name: TheOne
; Zone name: abc.com
; Date and time (UTC): Mon Jan 22 2018 09:09:43 GMT+0000

$TTL 3600
$ORIGIN abc.com.

@ 3600 IN SOA ns1-03.azure-dns.com. azuredns-hostmaster.microsoft.com. (
                                1
                                3600
                                300
                                2419200
                                300
                                )

@ 3600 IN A 1.1.1.1

@ 172800 IN NS ns1-03.azure-dns.com.
  172800 IN NS ns2-03.azure-dns.net.
  172800 IN NS ns3-03.azure-dns.org.
  172800 IN NS ns4-03.azure-dns.info.

www 3600 IN A 1.1.1.1
```

Import command:
```
~$ az network dns zone import -g dnsrepro -n abc.com --file-name /mnt/c/temp/azclidns/abc.txt
== BEGINNING ZONE IMPORT: abc.com ==

(4/7) Imported 4 records of type 'NS' and name '@'
(5/7) Imported 1 records of type 'soa' and name '@'
(6/7) Imported 1 records of type 'a' and name 'www.abc.com'
(7/7) Imported 1 records of type 'a' and name '@'

== 7/7 RECORDS IMPORTED SUCCESSFULLY: 'abc.com' ==
```

A record shows fqdn of www.abc.com.abc.com (expected just **www**):
```
~$ az network dns record-set list -g dnsrepro --zone-name abc.com
[
  {
    "arecords": [
      {
        "ipv4Address": "1.1.1.1"
      }
    ],
    "etag": "97ec82e6-3fa6-49db-8b29-621e9a4bceaa",
    "fqdn": "abc.com.",
    "id": "/subscriptions/2c8b66e1-8e65-44a9-872d-455bdb1b5184/resourceGroups/dnsrepro/providers/Microsoft.Network/dnszones/abc.com/A/@",
    "metadata": null,
    "name": "@",
    "resourceGroup": "dnsrepro",
    "ttl": 3600,
    "type": "Microsoft.Network/dnszones/A"
  },
  {
    "etag": "d92f5090-d4f6-4c98-acaa-b15e3e9b7f8c",
    "fqdn": "abc.com.",
    "id": "/subscriptions/2c8b66e1-8e65-44a9-872d-455bdb1b5184/resourceGroups/dnsrepro/providers/Microsoft.Network/dnszones/abc.com/NS/@",
    "metadata": null,
    "name": "@",
    "nsRecords": [
      {
        "nsdname": "ns1-09.azure-dns.com."
      },
      {
        "nsdname": "ns2-09.azure-dns.net."
      },
      {
        "nsdname": "ns3-09.azure-dns.org."
      },
      {
        "nsdname": "ns4-09.azure-dns.info."
      }
    ],
    "resourceGroup": "dnsrepro",
    "ttl": 172800,
    "type": "Microsoft.Network/dnszones/NS"
  },
  {
    "etag": "1b84e14c-65f9-4fbe-87a1-4575ad0d2692",
    "fqdn": "abc.com.",
    "id": "/subscriptions/2c8b66e1-8e65-44a9-872d-455bdb1b5184/resourceGroups/dnsrepro/providers/Microsoft.Network/dnszones/abc.com/SOA/@",
    "metadata": null,
    "name": "@",
    "resourceGroup": "dnsrepro",
    "soaRecord": {
      "email": "azuredns-hostmaster.microsoft.com.",
      "expireTime": 2419200,
      "host": "ns1-09.azure-dns.com.",
      "minimumTtl": 300,
      "refreshTime": 3600,
      "retryTime": 300,
      "serialNumber": 1
    },
    "ttl": 3600,
    "type": "Microsoft.Network/dnszones/SOA"
  },
  {
    "arecords": [
      {
        "ipv4Address": "1.1.1.1"
      }
    ],
    "etag": "80959454-3d66-4363-99b2-5d45b691be87",
    "fqdn": "www.abc.com.abc.com.",
    "id": "/subscriptions/2c8b66e1-8e65-44a9-872d-455bdb1b5184/resourceGroups/dnsrepro/providers/Microsoft.Network/dnszones/abc.com/A/www.abc.com",
    "metadata": null,
    "name": "www.abc.com",
    "resourceGroup": "dnsrepro",
    "ttl": 3600,
    "type": "Microsoft.Network/dnszones/A"
  }
]
```

With older versions (tested 2.0.20 and 2.0.24) it works as expected:

```
~$ az network dns record-set list -g dnsrepro --zone-name abc.com

== BEGINNING ZONE IMPORT: abc.com ==

(1/7) Imported 1 records of type 'soa' and name '@'
(2/7) Imported 1 records of type 'a' and name '@'
(6/7) Imported 4 records of type 'NS' and name '@'
(7/7) Imported 1 records of type 'a' and name 'www'

== 7/7 RECORDS IMPORTED SUCCESSFULLY: 'abc.com' ==

~$ az network dns record-set list -g dnsNorepro --zone-name abc.com
[
  {
    "arecords": [
      {
        "ipv4Address": "1.1.1.1"
      }
    ],
    "etag": "64d81fa4-85f0-4e57-8309-24d33eaa9abf",
    "id": "/subscriptions/2c8b66e1-8e65-44a9-872d-455bdb1b5184/resourceGroups/dnsnorepro/providers/Microsoft.Network/dnszones/abc.com/A/@",
    "metadata": null,
    "name": "@",
    "resourceGroup": "dnsnorepro",
    "ttl": 3600,
    "type": "Microsoft.Network/dnszones/A"
  },
  {
    "etag": "0315ea2d-91f1-4454-a38e-bfd70ec1f991",
    "id": "/subscriptions/2c8b66e1-8e65-44a9-872d-455bdb1b5184/resourceGroups/dnsnorepro/providers/Microsoft.Network/dnszones/abc.com/NS/@",
    "metadata": null,
    "name": "@",
    "nsRecords": [
      {
        "nsdname": "ns1-06.azure-dns.com."
      },
      {
        "nsdname": "ns2-06.azure-dns.net."
      },
      {
        "nsdname": "ns3-06.azure-dns.org."
      },
      {
        "nsdname": "ns4-06.azure-dns.info."
      }
    ],
    "resourceGroup": "dnsnorepro",
    "ttl": 172800,
    "type": "Microsoft.Network/dnszones/NS"
  },
  {
    "etag": "ca4dc8ad-fa6d-441f-919a-a3de802d1485",
    "id": "/subscriptions/2c8b66e1-8e65-44a9-872d-455bdb1b5184/resourceGroups/dnsnorepro/providers/Microsoft.Network/dnszones/abc.com/SOA/@",
    "metadata": null,
    "name": "@",
    "resourceGroup": "dnsnorepro",
    "soaRecord": {
      "email": "azuredns-hostmaster.microsoft.com.",
      "expireTime": 2419200,
      "host": "ns1-06.azure-dns.com.",
      "minimumTtl": 300,
      "refreshTime": 3600,
      "retryTime": 300,
      "serialNumber": 1
    },
    "ttl": 3600,
    "type": "Microsoft.Network/dnszones/SOA"
  },
  {
    "arecords": [
      {
        "ipv4Address": "1.1.1.1"
      }
    ],
    "etag": "b5154e76-431a-4de9-a1ec-60b06875b579",
    "id": "/subscriptions/2c8b66e1-8e65-44a9-872d-455bdb1b5184/resourceGroups/dnsnorepro/providers/Microsoft.Network/dnszones/abc.com/A/www",
    "metadata": null,
    "name": "www",
    "resourceGroup": "dnsnorepro",
    "ttl": 3600,
    "type": "Microsoft.Network/dnszones/A"
  }
]
```

### PR DESCRIPTION
**Pull request by tjprescott**
Fixes #5259. Fixes #5362.

Adds support for `--file-name` to support directly downloading to file. This was needed to make the test truly work (verify the import-export-import produces the same results) and will probably be useful in other scripting scenarios.

Additionally, this PR improves the testing of DNS zone files. Previously, we only tested the zone file parsing.  This was insufficient to catch a regression that occurred during the 2.0.24 release. Now, there is also a scenario test that verifies the following scenario for each of the test zone files: 

`import zone file > export zone > delete zone > re-import zone from exported zone file`

The records in the zones should be identical in whether the zone is imported from the original file or from an exported zone.

---

This checklist is used to make sure that common guidelines for a pull request are followed.

### General Guidelines

- [X] The PR has modified HISTORY.rst describing any customer-facing, functional changes. Note that this does not include changes only to help content. (see [Modifying change log](https://github.com/Azure/azure-cli/tree/master/doc/authoring_command_modules#modify-change-log)).

### Command Guidelines

- [N/A] Each command and parameter has a meaningful description.
- [N/A] Each new command has a test.

(see [Authoring Command Modules](https://github.com/Azure/azure-cli/tree/master/doc/authoring_command_modules))

### CODE REVIEW

**Comment by promptws** [comment]
View a preview at https://prompt.ws/r/Azure/azure-cli/5346
This is an **experimental preview** for @microsoft users.

File: src/command_modules/azure-cli-network/azure/cli/command_modules/network/custom.py [hunk 1]
```diff
@@ -966,7 +966,11 @@ def export_zone(cmd, resource_group_name, zone_name):
 
             zone_obj[record_set_name][record_type].append(record_obj)
 
-    print(make_zone_file(zone_obj))
+    zone_file = make_zone_file(zone_obj)
```
Comment (troydai) [inline src/command_modules/azure-cli-network/azure/cli/command_modules/network/custom.py]: A better variable name is `zone_file_content` to distinguish it from an actual file object, which is created in following code.

File: src/command_modules/azure-cli-network/azure/cli/command_modules/network/custom.py [hunk 2]
```diff
@@ -966,7 +966,11 @@ def export_zone(cmd, resource_group_name, zone_name):
...
+    print(zone_file)
+    if file_name:
+        with open(file_name, 'w') as f:
+            f.write(zone_file)
```
Comment (troydai) [inline src/command_modules/azure-cli-network/azure/cli/command_modules/network/custom.py]: I would recommend to capture `IOException` can translate it to a user-friendly error message.

File: src/command_modules/azure-cli-network/azure/cli/command_modules/network/custom.py [hunk 3]
```diff
@@ -900,7 +900,7 @@ def _type_to_property_name(key):
     return type_dict[key.lower()]
 
 
-def export_zone(cmd, resource_group_name, zone_name):
+def export_zone(cmd, resource_group_name, zone_name, file_name=None):
```
Comment (troydai) [inline src/command_modules/azure-cli-network/azure/cli/command_modules/network/custom.py]: do we need help description for the newly added argument?

Comment (tjprescott) [reply src/command_modules/azure-cli-network/azure/cli/command_modules/network/custom.py]: The help text is "Path to the DNS zone file to save."

**Comment by tjprescott** [comment]
@troydai comments addressed. Please take a second look.

**Pull request merged**